tests/Doctrine/Tests/ORM/Query/ExprTest.php

```diff
@@ -299,7 +299,7 @@ public function testNotInLiteralExpr()
 
     public function testNotInExprForEmptyArray()
     {
-        self::assertEquals('u.id NOT IN(FALSE)', (string) $this->_expr->notIn('u.id', []));
+        self::assertEquals('u.id NOT IN(NULL)', (string) $this->_expr->notIn('u.id', []));
```

Comment:
`u.id NOT IN(NULL)` would result in an empty result too.

https://dbfiddle.uk/?rdbms=sqlite_3.8&fiddle=0097a9450b4e56c142e64a480b745614

this contradicts with your expectations described in your issue https://github.com/doctrine/orm/issues/8033#issuecomment-589885694

Reply:
Yeah, I know.

It's still an improvement since 
- IN (NULL) works correclty
- NOT IN (NULL) wont return an error, when NOT IN () does.

And in the next comment https://github.com/doctrine/orm/issues/8033#issuecomment-589950348, I explained that
```
$query->andWhere("foo.bar NOT IN (:baz)")->setParameter('baz', []);
```
Is actually translated as `foo.bar NOT IN (NULL)`.

So my PR make the behaviour of `$query->expr()->notIn()` more consistent with `$query->andWhere("foo.bar NOT IN (:baz)")->setParameter('baz', []);`.

Reply:
We can improve the behaviour of the NOT IN query with an empty array, but It'll need a lot a more code changes and some discussions, so I recommend to split this from this PR, which was only looking for more consistency.

Reply:
If you want to split it, then you'll have to remove the `NOT IN` changes. What do you want to change for supporting this case?

Reply:
> If you want to split it, then you'll have to remove the `NOT IN` changes. 

I disagree. Actually if you write
```
$query->andWhere($queryBuilder->expr()->notIn('foo.bar', ':baz'))
$query->setParameter('baz', []);
```
There is no exception and this is translated as `foo.bar NOT IN (NULL)`.

But if you write directly
```
$query->andWhere($queryBuilder->expr()->notIn('contractOperation.type', []))
```
You get an exception. But there is absolutely no reason to behave differently from the first case.

So it's still an improvement to have a consistent behaviour for `notIn('foo.bar', ':baz'))` and `notIn('foo.bar', []))`.

I improve both the `IN` case and the `NOT IN` case.
So I don't think I should remove the `NOT IN` changes. 

I don't know what I have to change to completely supports the NOT IN case and if it's possible. It should handle correctly the case 
```
$query->andWhere('foo.bar NOT IN :baz')
$query->setParameter('baz', []);
```
```
$query->andWhere($queryBuilder->expr()->notIn('foo.bar', ':baz'))
$query->setParameter('baz', []);
```
```
$query->andWhere($queryBuilder->expr()->notIn('foo.bar', []))
```
And this seems for me like a lot of work/rework.

Is this PR perfect and fixing everything ? No
Is this PR better than the actual situation ? Yes